File: practica_8/Main.java
/*En java no hay herencia multiple de forma nativa pero se puede simular con interfases  y compocision
en este caso voy a usar compocicion para que se parezca a el diagrama*/

public class A {
    private int x;
    private int z;

    public A(int x, int z) {
        this.x = x;
        this.z = z;
    }

    public void incrementaXZ() {
        x++;
        z++;
    }

    public void incrementaZ() {
        z++;
    }

    public int getX() {
        return x;
    }

    public int getZ() {
        return z;
    }
}

public class B {
    private int y;
    private int z;

    public B(int y, int z) {
        this.y = y;
        this.z = z;
    }

    public void incrementaYZ() {
        y++;
        z++;
    }

    public void incrementaZ() {
        z++;
    }

    public int getY() {
        return y;
    }

    public int getZ() {
        return z;
    }
}

public class D {
    private A a;
    private B b;

    public D(int x, int y, int z) {
        this.a = new A(x, z);
        this.b = new B(y, z);
    }

    public void incrementaXZ() {
        a.incrementaXZ();
    }

    public void incrementaYZ() {
        b.incrementaYZ();
    }

    public void incrementaZ() {
        a.incrementaZ();
        b.incrementaZ();
    }

    public void incrementaXYZ() {
        a.incrementaXZ();
        b.incrementaYZ();
    }
}

File: practica_8/main.py
class A:
    def __init__(self, x, z):
        self.x = x
        self.z = z

    def incrementaXZ(self):
        self.x += 1
        self.z += 1

    def incrementaZ(self):
        self.z += 1

class B:
    def __init__(self, y, z):
        self.y = y
        self.z = z

    def incrementaYZ(self):
        self.y += 1
        self.z += 1

    def incrementaZ(self):
        self.z += 1

class D(A, B):
    def __init__(self, x, y, z):
        A.__init__(self, x, z)
        B.__init__(self, y, z)

    def incrementaXYZ(self):
        self.x += 1
        self.y += 1
        self.z += 1

    def incrementaZ(self):
        A.incrementaZ(self)
        B.incrementaZ(self)
        
